### PR TITLE
Generator enhancements

### LINKED ISSUE #651: [Bug]: Product generator not filling the product block relations table
### Contact Details

_No response_

### What happened?

The product generator is not filling the mapper when it is generating migrations correctly. The dependancy tree on the product side is nog being filled out correctly. This is a minor issue currently as the domain models work correctly, however if we use this table more in the future it would be nice if the generator implements this best practice.

### Version

2.3.0rc3

### What python version are you seeing the problem on?

All

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Mark90**
## Generator enhancements

**product_block**
- Fix imports of dependent product blocks in the same product
- Detect invalid product block configuration: cyclic or multiple roots

**product**
- Only attach the root product block to the product type

**migration**
- Only relate the root product block to the product, instead of all blocks (`product_product_blocks`)
- Fill depends_on_block_relations, ensuring the correct order (`product_block_relations`)
- Only create resource types for non-product block fields (`product_block_resource_types`)

Please review the updated migrations in the diff, as well as in the example orchestrator https://github.com/workfloworchestrator/example-orchestrator/pull/19

## Other

- add testing for generated migrations
- add testcases for (invalid) product block generation
- make revision argument to command `db upgrade` mandatory
- change `db history` command to indicate current revision by default

Closes #651

### CODE REVIEW

**Review by pboers1988 — APPROVED** [review]

File: orchestrator/cli/generator/generator/helpers.py [hunk 1]
```diff
@@ -44,10 +49,117 @@ def get_product_block_file_name(product_block: dict) -> str:
...
+
+    block_order: dict[str, int] = {}
+    order = 0
+    while block_dependencies:
+        cycle = True
+
+        for block, depends_on_blocks in list(block_dependencies.items()):
+            if depends_on_blocks - block_order.keys():
+                # Not all dependent blocks are resolved yet
+                continue
```
Comment (mrijk) [inline orchestrator/cli/generator/generator/helpers.py:136]: When I see 'continue' and 'del' I start thinking about poor baby seals

**Review by mrijk — APPROVED** [review]
Bit too many nested for loops (and if statements) for my taste

File: orchestrator/cli/generator/generator/migration.py [hunk 2]
```diff
@@ -134,25 +140,39 @@ def generate_product_migration(context: dict) -> None:
...
-    else:
+    if not (migration_file := create_migration_file(message=f"add {config['name']}", head="data")):
         logger.error("Could not create migration file")
+        return
+
+    if not (revision_info := get_revision_info(migration_file)):
+        logger.error("Could not get revision info from migration file", migration_file=migration_file)
+        return
+
+    if "fixed_inputs" in config and config["fixed_inputs"]:
```
Comment (tjeerddie) [inline orchestrator/cli/generator/generator/migration.py:151]: use `.get()` with walrus operator
```suggestion
    if fixed_inputs := config.get("fixed_inputs"):
```

File: orchestrator/cli/generator/generator/migration.py [hunk 3]
```diff
@@ -134,25 +140,39 @@ def generate_product_migration(context: dict) -> None:
...
+        fixed_input_values = [
+            [(fixed_input["name"], str(value)) for value in fixed_input["values"]]
+            for fixed_input in config["fixed_inputs"]
```
Comment (tjeerddie) [inline orchestrator/cli/generator/generator/migration.py:154]: to batch with above comment
```suggestion
            for fixed_input in fixed_inputs
```

File: orchestrator/cli/generator/generator/helpers.py [hunk 4]
```diff
@@ -44,10 +49,117 @@ def get_product_block_file_name(product_block: dict) -> str:
     return get_product_block_variable(product_block)
 
 
+def get_existing_product_blocks() -> dict[str, Any]:
```
Comment (tjeerddie) [inline orchestrator/cli/generator/generator/helpers.py:52]: move product block helpers to its own file for shorter files and a bit more clean

**Review by tjeerddie — APPROVED** [review]

File: orchestrator/cli/generator/generator/helpers.py [hunk 5]
```diff
@@ -44,10 +49,117 @@ def get_product_block_file_name(product_block: dict) -> str:
...
+            name = pb_file.removesuffix(".py")
+            module_name = f"{get_product_blocks_module()}.{name}"
+
+            module = import_module(module_name)
+
+            classes = [obj for _, obj in inspect.getmembers(module, inspect.isclass) if obj.__module__ == module_name]
+
+            for klass in classes:
+                if is_product_block(klass):
+                    yield klass.__name__, module_name
```
[the remark this answers is not in the document]
Comment (Mark90) [reply orchestrator/cli/generator/generator/helpers.py:73]: I merely moved this function (git says it was yours actually) but I'll change it

[the remark this answers is not in the document]
Comment (mrijk) [reply orchestrator/cli/generator/generator/helpers.py]: Let's just say I got wiser over the years ;)

**Pull request merged**